File: python/util_py/prime.py
#! /usr/bin/python
from math import sqrt, ceil

def is_prime(n):
    if n==1: return False
    if n==2: return True
    for i in range(2, int(ceil(sqrt(n)))+1):
        if n%i==0:
            return False
    return True

def primes_up_to(n):
    primes = []
    flag = False
    for i in range(2,n):
        for p in sorted(primes):
            if p <= int(ceil(sqrt(i))):
                if i % p == 0:
                    flag = True
                    break
        if not flag: primes.append(i)
        flag = False
    return primes

File: ruby/util/prime.rb
#!/usr/bin/env ruby

def sieve(max)
	n_sq = max
	n = Math.sqrt(n_sq)
	a =(0..n_sq).to_a
	a[0] = nil
	a[1] = nil
	index = 2

	while index <= n do
		# remove multiples of index up to n_sq
		multiple = index * 2
		while multiple <= n_sq do
			a[multiple] = nil
			multiple += index
		end
		
		# get next index
		index += 1
		while index <=n and a[index] == nil do
			index += 1
		end
	end
	s = SortedSet.new
	s.merge(a.compact)
end

def prime_generator()
	primes = [2]
	yield 2
	candidate = 3
	while true
		is_prime = true
		upper_bound = (Math.sqrt(candidate)).floor
		for prime in primes
			if prime > upper_bound
				break
			end
			if (candidate % prime).zero?
				is_prime = false
				break
			end
		end
		if is_prime
			primes << candidate
			yield candidate
		end
		candidate += 2
	end
end
